lib/Runtime/Library/JavascriptSet.cpp

```diff
@@ -6,379 +6,669 @@
+    if (!JavascriptSet::Is(args[0]))
+    {
+        JavascriptError::ThrowTypeErrorVar(scriptContext, JSERR_NeedObjectOfType, _u("Set.prototype.entries"), _u("Set"));
     }
 
-    Var JavascriptSet::EntryForEach(RecyclableObject* function, CallInfo callInfo, ...)
-    {
-        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+    JavascriptSet* set = JavascriptSet::FromVar(args[0]);
 
-        ARGUMENTS(args, callInfo);
-        ScriptContext* scriptContext = function->GetScriptContext();
-        AUTO_TAG_NATIVE_LIBRARY_ENTRY(function, callInfo, _u("Set.prototype.forEach"));
+    return scriptContext->GetLibrary()->CreateSetIterator(set, JavascriptSetIteratorKind::KeyAndValue);
+}
 
-        if (!JavascriptSet::Is(args[0]))
-        {
-            JavascriptError::ThrowTypeErrorVar(scriptContext, JSERR_NeedObjectOfType, _u("Set.prototype.forEach"), _u("Set"));
-        }
+Var JavascriptSet::EntryValues(RecyclableObject* function, CallInfo callInfo, ...)
+{
+    PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
 
-        JavascriptSet* set = JavascriptSet::FromVar(args[0]);
+    ARGUMENTS(args, callInfo);
+    ScriptContext* scriptContext = function->GetScriptContext();
 
+    if (!JavascriptSet::Is(args[0]))
+    {
+        JavascriptError::ThrowTypeErrorVar(scriptContext, JSERR_NeedObjectOfType, _u("Set.prototype.values"), _u("Set"));
+    }
 
-        if (args.Info.Count < 2 || !JavascriptConversion::IsCallable(args[1]))
-        {
-            JavascriptError::ThrowTypeError(scriptContext, JSERR_FunctionArgument_NeedFunction, _u("Set.prototype.forEach"));
-        }
-        RecyclableObject* callBackFn = RecyclableObject::FromVar(args[1]);
+    JavascriptSet* set = JavascriptSet::FromVar(args[0]);
 
-        Var thisArg = (args.Info.Count > 2) ? args[2] : scriptContext->GetLibrary()->GetUndefined();
+    return scriptContext->GetLibrary()->CreateSetIterator(set, JavascriptSetIteratorKind::Value);
+}
 
-        auto iterator = set->GetIterator();
+Var JavascriptSet::EntryGetterSymbolSpecies(RecyclableObject* function, CallInfo callInfo, ...)
+{
+    ARGUMENTS(args, callInfo);
 
-        while (iterator.Next())
-        {
-            Var value = iterator.Current();
+    Assert(args.Info.Count > 0);
 
-            CALL_FUNCTION(scriptContext->GetThreadContext(), callBackFn, CallInfo(CallFlags_Value, 4), thisArg, value, value, args[0]);
-        }
+    return args[0];
+}
 
-        return scriptContext->GetLibrary()->GetUndefined();
-    }
+template <typename T>
+T*
+JavascriptSet::CreateVarSetFromList(uint initialCapacity)
+{
+    T* varSet = RecyclerNew(GetRecycler(), T, GetRecycler(), initialCapacity);
 
-    Var JavascriptSet::EntryHas(RecyclableObject* function, CallInfo callInfo, ...)
+    JavascriptSet::SetDataList::Iterator iter = this->list.GetIterator();
+    // TODO: we can use a more efficient Iterator, since we know there will be no side effects
```

Comment:
this is a good point. Should we mark as JS_NO_REENTRANT region so that it is easy to explain that?